Post: NServiceBus.IntegrationTesting baby steps (#228)
mauroservienti/milestone.github.io: _posts/nservicebus-integrationtesting-baby-steps.md

---
layout: post
header_image: /img/posts/nservicebus-integrationtesting-baby-steps/header.jpg
title: "NServiceBus.IntegrationTesting baby steps"
author: Mauro Servienti
synopsis: "NServiceBus.IntegrationTesting started as an exploration activity and recently turned into a stable v1 release. V1 comes with some new features and one limitation that led to brainstorming options for the future."
tags:
- NServiceBus
- Integration testing
- SOA
---

What started as an experiment evolved a month ago into a stable release. [NServiceBus.IntegrationTesting](https://github.com/mauroservienti/NServiceBus.IntegrationTesting) v1 is available on [NuGet](https://www.nuget.org/packages/NServiceBus.IntegrationTesting/).

The integration testing framework saw the light mainly as an experiment; you can read more about it in ["Exploring NServiceBus integration testing options"](https://milestone.topics.it/2019/07/04/exploring-nservicebus-integration-testing-options.html). The idea was to explore the problem space first. The main questions were:

- What kind of integration testing are NServiceBuse users expecting to perform?
- Which parts of the system cannot be unit tested?

> As mentioned in the above-linked article, the main driver is what I like to call choreography testing when it comes to integration testing. Choreography testing's primary goal is to validate that one or more business scenarios behave as expected. A business scenario is handled by many components in a distributed system, being autonomous or business components, hosted by as many again endpoints. The sample provided in the GitHub repository goes like follows:

Reviewer: ```suggestion
As mentioned in the above article, the main driver is what I like to call choreography testing when it comes to integration testing. Choreography testing's primary goal is to validate that one or more business scenarios behave as expected. A business scenario is handled by many components in a distributed system, being autonomous or business components, hosted by as many endpoints. The sample provided in the GitHub repository goes like this:
```

- `MyService` sends `AMessage` to `MyOtherService`.
- `MyOtherService` replies with a `AReplyMessage`.
- `MyService` handles `AReplyMessage` and sends locally to itself a `StartASaga` message.
- `ASaga` is started in `MyService`.

> Looking ab the above "business story," the test needs to be able to assert that `ASaga` started and that `MyService` handled `AReplyMessage`.

Reviewer: ```suggestion
Looking at the above "business story," the test needs to be able to assert that `ASaga` started and that `MyService` handled `AReplyMessage`.
```

- A set of libraries to execute tests and connect the dots between the tests and the chosen test runner (as of today, you're forced to use NUnit)
- A Docker container base image to host each endpoint under test in isolation and a set of tools to communicate with the libraries in the test agent process

With that in place, we could write something like:

```csharp
public async Task Pseudo_test_this_is_hypothetical()
{
   var context = new IntegrationTestContext();
   var scenario = new IntegrationScenario(context);
   
   //relative path to C# project
   scenario.AddEndpoint("../../MyEndpoint", onStarted: session => { ... });
   scenario.AddEndpoint("../../MyOtherEndpoint");
   
   scenario.Done(c => c.MessageWasProcessedByHandler(...) || c.HasFailedMessages );

   await scenario.Run();

   Assert.True(...);
}
```

> When run, the above scenario packages the supplied C# projects into containers and starts them. Each container will communicate with the test host, e.g., via HTTP, keep the host informed about execution status, and report handled messages and executed sagas. The host will then evaluate the supplied done conditions and determine when the test execution is complete.

Reviewer: ```suggestion
When run, the above scenario packages the supplied C# projects into containers and starts them. Each container will communicate with the test host, e.g. via HTTP, keep the host informed about execution status, and report handled messages and executed sagas. The host will then evaluate the supplied done conditions and determine when the test execution is complete.
```
